lightningd/pay.c

```diff
@@ -1445,14 +1547,26 @@ static struct command_result *json_sendpay(struct command *cmd,
 		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
 				    "Must specify msatoshi with partid");
 
-	const struct amount_msat final_amount = route[tal_count(route)-1].amount;
-
 	/* If groupid was not provided default to incrementing from the previous one. */
 	if (group == NULL) {
 		group = tal(tmpctx, u64);
 		*group = wallet_payment_get_groupid(cmd->ld->wallet, rhash) + 1;
 	}
 
+	if (tal_count(route) == 0) {
+		if (!msat)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "Self-payment requires an amount");
+		if (*partid)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "Self-payment does not allow (non-zero) partid");
+		return self_payment(cmd->ld, cmd, rhash, *partid, *group, *msat,
+				    label, invstring, description, local_invreq_id,
+				    payment_secret, payment_metadata);
+	}
+
+	const struct amount_msat final_amount = route[tal_count(route)-1].amount;
+
 	if (msat && !*partid && !amount_msat_eq(*msat, final_amount))
 		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
 				    "Do not specify msatoshi (%s) without"
```

Comment:
Ok I am not able to check the sanity of the following case, what happens if:

Node A generates an invoice where that contains a route where the last hop is the node that it is paying the invoice.

I guess this is the use case of self_payment for fedimint

Reply:
> Node A generates an invoice where that contains a route where the last hop is the node that it is paying the invoice.

Correct. Phrased differently, _Fedimint self-payment arises if one client (alice) creates an invoice and delegates it's receive to a given gateway node G. Another client (bob) instructs G to pay the invoice. Thus both  alice and bob have delegated the lightning pay and receive to G_

Reply:
> Ok I am not able to check the sanity of the following case, what happens if:
> 
> Node A generates an invoice where that contains a route where the last hop is the node that it is paying the invoice.
> 
> I guess this is the use case of self_payment for fedimint

We haven't even looked at routes at this point, so it should Just Work.